common/scala/src/main/scala/whisk/core/database/cosmosdb/CosmosDBSupport.scala

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package whisk.core.database.cosmosdb
+
+import com.microsoft.azure.cosmosdb._
+import com.microsoft.azure.cosmosdb.rx.AsyncDocumentClient
+
+import scala.collection.JavaConverters._
+import scala.collection.immutable
+
+private[cosmosdb] trait CosmosDBSupport extends RxObservableImplicits with CosmosDBUtil {
+  protected def config: CosmosDBConfig
+  protected def collName: String
+  protected def client: AsyncDocumentClient
+  protected def viewMapper: CosmosDBViewMapper
+
+  def initialize(): (Database, DocumentCollection) = {
+    val db = getOrCreateDatabase()
+    (db, getOrCreateCollection(db))
+  }
+
+  private def getOrCreateDatabase(): Database = {
+    client
+      .queryDatabases(querySpec(config.db), null)
```

Comment:
is it possible to avoid `null` and use None (viz Option) instead?
we've generally stayed away from null.

Reply:
`queryDatabases` is java api of Azure CosmosDB driver which uses the `null` based approach. So not possible to use Option here

Reply:
> getOrCreateReference is called once at create time

Added some comments. This method may be called concurrently multiple times at startup as various store instances are initialized. Hence need to `synchronize` it to ensure all those stores are backed by same client